### ISSUE TITLE
local_engine uploads in the flutter_deps recipe module will fail with isolate -> CAS migration

### ISSUE THREAD

**Issue by whesse** [flutter/flutter#84119]
The isolate servers are deprecated, and will be shut down, and are replaced by CAS servers.  Flutter recipes are migrated to use CAS for most of their uses, because the use_cas property is passed to swarming.

But the special case where the flutter_deps recipe module can download a local engine build, and pass the LOCAL_ENGINE environment variable to all the tools run in the recipe, is still using the isolate server:
https://flutter.googlesource.com/recipes/+/refs/heads/master/recipe_modules/flutter_deps/api.py#29

This server isolateserver.appspot.com is replaced by projects/chromium-swarm/instances/default_instance.
In the recipe, the use of the "isolated" module should be replaced by a use of the "cas" module or of the "build/isolate" module, according the the migration instructions.

A problem arises when migrating to RBE-CAS because uploads of directories to CAS do not allow symbolic links, and the flutter engine builds have symlinks in the out/[configuration]/gen/dart-pkg/packages directory from sky_engine to ../sky_engine/lib and from sky_services to ../sky_services/lib.

Perhaps we can copy the files, or make hard links, or move them, to avoid this problem.  About 15 symlinks are shown if I search the directory tree with "ls -lR out/android_release_arm64/ | grep ^l".

I am using this local_engine upload in the flutter_deps module to run triple-headed (tip-of-tree of Dart, Engine, and Framework) builds of Flutter on the device lab for performance testing, which we want to be able to spot performance changes due to Dart SDK commits.

@CaseyHillers @godofredoc @sortie @athomas

**Comment by whesse** [comment]
The symbolic links that are the problem are links to outside the build directory, and they are only in two directories.
The fix (and also a hack in a script before the fix is landed) is really simple: out/[configuration]/gen/dart-pkg/sky_engine and sky_services should just be copied, rather than linked to the source tree. The directories are only 10 MB and 5 MB.

So this fix (of not linking these directories) should be landed, I think, and we will do the workaround until then.

**Comment by whesse** [comment]
(And this CAS migration isn't being forced soon, and nothing is broken until then.  We just didn't want to use isolate instead of CAS when using the local recipes in a flutter recipe (flutter_deps flutter_engine() function) that currently uses isolate.

**Comment by whesse** [comment]
Created issue #84649 to remove isolate support from the flutter_deps recipe for local engine.

**Comment by whesse** [comment]
CAS support is added to flutter_deps in the CL https://flutter-review.googlesource.com/c/recipes/+/14640
Systems using this CAS support are working around this issue with symbolic links by manually replacing them with copies using scripts, until this issue is fixed.

**Comment by zanderso** [comment]
I would like an answer to my question in https://github.com/flutter/flutter/issues/84119#issuecomment-857256867. I am concerned that a workaround in infra will make it difficult to land the real fix to the Engine build. If this is urgent, then please pursue the fix to the Engine build instead of working around the symlinks in the infra scripts.

**Comment by hamad22h** [comment]
I

**Comment by whesse** [comment]
I don't know the timeline for shutting down the servers - I think it is months.  I would like to get the build fixed quickly, to have no symbolic links, though, just because it is needed relatively soon. Relative symbolic links within a single build directory are not a problem for us, though.

Our fix just copies the directories sky_engine and sky_services containing symbolic links, using options that replace the links with their contents.   The copied directories are then moved to the original locations, after deleting the originals.

**Comment by godofredoc** [comment]
@whesse everything has been migrated to CAS can we close this bug?

**Comment by whesse** [comment]
I'd like to do a second pass, to see if anything else was using the local_engine upload in the flutter_deps recipe module.  If those uses have been migrated to use cas (the new local_engine_cas_hash input property), they must be doing something to make their uploads work.
If they haven't migrated, then not everything has migrated to CAS.  So I think this issue should be open, until we have verified that the links to sky_engine are no longer symbolic links.

I'll investigate these issues, but someone more familiar with all recipes used in Flutter Luci would be better suited to it.  I haven't heard that the symbolic links in the build have been removed, and that our workaround can be removed.
